### PR TITLE
Added new features of TouchPortal API

### PR DESCRIPTION
**Pull request by KillerBOSS2019**
Updated class TYPES Notification events
Added showNotification method
Added connectorUpdate method

### CODE REVIEW

**Comment by KillerBOSS2019** [comment]
I do not have TouchPortal 2.4 to test this

File: TouchPortalAPI/__init__.py [hunk 1]
```diff
@@ -286,6 +324,33 @@ def stateUpdateMany(self, states:list):
...
+
+    def connectorUpdate(self, connectorId:str, connectorValue:int):
+        if isinstance(connectorId, str):
+            if isinstance(connectorValue, int):
+                if 0 <= connectorValue <= 100:
+                    pass
+                else:
+                    raise TypeError(f"connectorValue needs to be between 0-100 not {connectorValue}")
+            else:
+                raise TypeError(f"connectorValue require a int not {type(connectorValue)}")
```
Comment (Pjiesco) [inline TouchPortalAPI/__init__.py:350]: `require` >> `requires`

File: TouchPortalAPI/__init__.py [hunk 2]
```diff
@@ -211,12 +213,21 @@ def __getWriteLock(self):
...
 
     def isActionBeingHeld(self, actionId:str):
+        '''
+        This returns True/False if the action id that you've passed in is pressed down or not
+        '''
         return actionId in self.__heldActions
 
     def createState(self, stateId:str, description:str, value:str):
+        '''
+        createState allows you to create a new states in runtime
```
Comment (Pjiesco) [inline TouchPortalAPI/__init__.py:229]: `in runtime` >> `at runtime`

File: TouchPortalAPI/__init__.py [hunk 3]
```diff
@@ -235,39 +250,59 @@ def createStateMany(self, states:list):
...
+        settingUpdate(settingName:str, settingValue) allows you to update
+        a specific setting fields with a new value
+        '''
         if settingName and settingName not in self.currentSettings or self.currentSettings[settingName] != settingValue:
             self.send({"type": "settingUpdate", "name": settingName, "value": settingValue})
             self.currentSettings[settingName] = settingValue
 
     def stateUpdate(self, stateId:str, stateValue:str):
+        '''
+        This allow existing states to update with a new value
```
Comment (Pjiesco) [inline TouchPortalAPI/__init__.py:304]: `allow` >> `allows`

**Review by Pjiesco — CHANGES_REQUESTED** [review]
Just some small typo's

File: TouchPortalAPI/__init__.py (hunk 1 above)
Comment (Pjiesco) [inline TouchPortalAPI/__init__.py:350]: does this do anything?

Comment (KillerBOSS2019) [reply TouchPortalAPI/__init__.py:350]: Woops I think I forgot that part.... Thanks

**Comment by KillerBOSS2019** [comment]
All the typos should be fixed now and the connectorUpdate. Now just need to wait for @mpaperno to double check this

**Review by mpaperno — APPROVED** [review]
Pushed a fix and some cosmetics. Otherwise looks good.   👍🏼 

Also +1 for adding docstrings to the methods.  Is there any particular reason you didn't use the same text as is in the README? I had gone through at one point and edited all those.

On a related note, I've been playing with generating docs automatically (using `pdoc` module), and that's pretty much good to go. So ultimately I'd suggest all the actual API docs should be in the code (docstrings) and should be auto-generated whenever there are changes. This will avoid duplicating docs into the README, and provide a real reference.

-Max

**Pull request merged**